### PR TITLE
chore: create docker image and compose for dev using the watch feature

### LINKED ISSUE #2: Ajout environnents dev et prod au Docker ?
Ça serait cool de pas devoir rebuild le contianer à chaque modification dans les projets Web et API.

Des idées ?

### PR DESCRIPTION
**Pull request by pbrissaud**
This commit create new image and compose file for dev mode. It use new compose feature `watch` released in version **2.22.0 (september 2023)** to automatically reload container when changes are detected

Video demo (password: tamagokill)
https://vimeo.com/952926923?share=copy

There is just a small delay for the API after the change; the time it takes to rebuild.

Don't forget `--watch` flag in docker compose command to enable this.

Example : `docker compose -f compose.dev.yaml up --build --watch`

Resources : 
- Cargo watch : https://crates.io/crates/cargo-watch
- Docker compose watch : https://docs.docker.com/compose/file-watch/

### CODE REVIEW

File: compose.yaml [hunk 1]
```diff
@@ -6,23 +6,13 @@ services:
...
-        protocol: tcp
-      # HTTPS
-      - target: 443
-        published: ${HTTPS_PORT:-443}
-        protocol: tcp
-      # HTTP/3
-      - target: 443
-        published: ${HTTP3_PORT:-443}
-        protocol: udp
+      - 8080:8080
```
Comment (lilymonad) [inline compose.yaml]: Revert this change, it is totaly OK to have it since the API could be exposed to outside.
If you want to modify default port forwarding just use environment variables on your local configuration like so:

```
HTTP_PORT=54321 docker-compose up
```
(or any other values of HTTP/HTTPS/HTTP3 ports you want)


It would also be nice to have it as an environment variable for the `web` container.

File: web/package.json [hunk 2]
```diff
@@ -36,5 +36,10 @@
 	"type": "module",
 	"dependencies": {
 		"dotenv": "^16.4.5"
-	}
+	},
+	"overrides": { 
+    "vite": {
+      "rollup": "npm:@rollup/wasm-node"
+    }
+  }
```
Comment (lilymonad) [inline web/package.json]: I don't understand why re-installing vite.
Didn't we agreed on using svelte @Lauwed ?

File: compose.dev.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        - action: sync
+          path: ./api
+          target: /app
+        - action: rebuild
+          path: ./api/Cargo.toml
+    restart: unless-stopped
+    depends_on:
+      - database
+    ports:
+      - 8080:8080
```
Comment (lilymonad) [inline compose.dev.yaml]: Same comment as for the compose.yam file, just use the same port forwarding layout, and if you want specific ports just use environment variables

File: compose.dev.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      watch:
+        - action: sync
+          path: ./web
+          target: /app
+          ignore:
+            - node_modules/
+        - action: rebuild
+          path: ./web/package.json
+    ports: 
+      - 3000:5173
```
Comment (lilymonad) [inline compose.dev.yaml]: use environment variables

**Review by lilymonad — CHANGES_REQUESTED** [review]

File: api/Dockerfile.dev [hunk 5]
```diff
@@ -0,0 +1,13 @@
+ARG RUST_VERSION=1.78.0
+
+FROM rust:${RUST_VERSION}-slim-bookworm
+
+WORKDIR /app
+
+COPY . .
```
Comment (lilymonad) [inline api/Dockerfile.dev:7]: Prefer binding at compose.dev.yaml level instead of copying the workspace in the development image.

Also why `cargo install --path .` since it's a ""devcontainer"" ?

File: web/package.json (hunk 2 above)
Comment (pbrissaud) [reply web/package.json]: Had a problem when installing the web dependencies on my machine (MacOS) . Found a similar issue and workaround here: https://github.com/vitejs/vite/issues/15167#issuecomment-1833274104

Comment (Lauwed) [reply web/package.json]: I didn't take the time to read anything right now, but I know that SvelteKit is build on Vite.

File: compose.yaml (hunk 1 above)
Comment (pbrissaud) [reply compose.yaml]: The API doesn't listen to 443 and 80 so the declaration of these ports are useless  https://github.com/DevGirl-Team/tamagokill/blob/665d9616757c4bce55fe4bbbd7237de688234830/api/src/main.rs#L22

Ok for the local port configuration as env 👍 Will change the code

File: api/Dockerfile.dev (hunk 5 above)
Comment (pbrissaud) [reply api/Dockerfile.dev:7]: > Prefer binding at compose.dev.yaml level instead of copying the workspace in the development image.

The watch feature in Docker Compose is designed to avoid volume binding due to its performance issues (source: [Docker Blog](https://www.docker.com/blog/announcing-docker-compose-watch-ga-release/)). I've taken the example of the compose documentation page : https://docs.docker.com/compose/file-watch/ 

> Also why cargo install --path . since it's a ""devcontainer"" ?

Even in a development container, we need to install the cargo dependencies. Running `cargo install` results in an error that can be resolved by using `cargo install --path .`
However, as I'm not a Rust expert, there might be a more optimal approach.

File: compose.yaml (hunk 1 above)
Comment (lilymonad) [reply compose.yaml]: Agreed on the TLS/HTTP3 ports **but** I think it's not that problematic to keep them for now, since it makes sense for the future.

File: api/Dockerfile.dev (hunk 5 above)
Comment (lilymonad) [reply api/Dockerfile.dev:7]: Fair enough, I've got poor windows dev experience I did not see the WSL problem. Thanks for documenting it !

PS: Man it's insane, an entire layer rebuild is more performant than binding files through a vm ? ...

**Comment by pbrissaud** [comment]
Added  2 new env for compose `API_PORT` and `WEB_PORT` to define the local port of these services

File: web/package.json (hunk 2 above)
Comment (pbrissaud) [reply web/package.json]: I can delete this change for this PR but I think other contributor on Mac could have this issue

File: api/Dockerfile.dev (hunk 5 above)
Comment (lilymonad) [reply api/Dockerfile.dev:7]: > Even in a development container, we need to install the cargo dependencies. 

Well `cargo install` is for installing the built executable somewhere, it is just doing `cargo build` + copying artefacts.

Same as `cargo run`, it will `cargo build` before running the executable if it was not previously built.
A simple `cargo build` should suffice, and I'm not even sure why you need to build it before `cargo watch`

Comment (pbrissaud) [reply api/Dockerfile.dev:7]: Maybe you're right on this ! Will test on my side

**Comment by lilymonad** [comment]
Ach too bad I dumped docker for podman, I can't test the feature (no --watch for podman-compose) ... well I guess a nix devShell wouldn't be a bad thing then lol silly me :p

**Comment by pbrissaud** [comment]
Both solutions can very well coexist to include as many people as possible!

**Comment by vasilvestre** [comment]
Why would you use a Dockerfile.dev ?
Multistage is made to avoid having multiple files and would be great to use here

**Comment by pbrissaud** [comment]
Using a multistage build is a viable approach, but having a separate Dockerfile.dev significantly enhances the readability and maintainability of our Docker setup. The distinct files allow developers to differentiate between development and production environments, making it easier to understand and modify the build process for each use case.

**Comment by pbrissaud** [comment]
Some fixes :
- Fix the default port for the web app
- Remove useless instructions on Dockerfile.dev (ty @lilymonad for the Rust side)
- Retrigger build when dockerfile changes

File: web/Dockerfile.dev [hunk 6]
```diff
@@ -0,0 +1,13 @@
+FROM node:18-alpine
+
+WORKDIR /app
+
+COPY package*.json .
+
+RUN npm ci
```
Comment (vasilvestre) [inline web/Dockerfile.dev]: You can add --audit false and --fund false for extra performance and readability

File: compose.yaml [hunk 7]
```diff
@@ -6,23 +6,13 @@ services:
...
-      - target: 443
-        published: ${HTTP3_PORT:-443}
-        protocol: udp
+      - ${API_PORT:-8080}:8080
 
   web:
     build: ./web
     ports: 
-      - 3000:3000
+     - ${WEB_PORT:-8080}::3000
```
Comment (ferdodo) [inline compose.yaml]: there's a double colon here, I think you need only one

Comment (pbrissaud) [reply compose.yaml]: Thanks you !

File: web/package.json (hunk 2 above)
Comment (ferdodo) [reply web/package.json]: @pbrissaud can you still reproduce the problem with an up to date version of sveltekit ?

Comment (pbrissaud) [reply web/package.json]: @ferdodo Nope, I can create a new svelte-kit project on my Mac without any issues. 

And now, there are no more errors on this project when I delete the "overrides" part... I will delete it!

Comment (ferdodo) [reply web/package.json]: @pbrissaud  perfect ! 🎉

**Review by vasilvestre — APPROVED** [review]

File: web/Dockerfile.dev [hunk 8]
```diff
@@ -0,0 +1,13 @@
+FROM node:18-alpine
```
Comment (ferdodo) [inline web/Dockerfile.dev]: Can you also define a environment variable for node version ?

this could be useful as mentioned here https://github.com/DevGirl-Team/tamagokill/pull/7#discussion_r1629225992

Comment (pbrissaud) [reply web/Dockerfile.dev]: Of course! In dev and prod images?

Comment (vasilvestre) [reply web/Dockerfile.dev]: You should run the same version for dev and prod. Take a look at https://github.com/DevGirl-Team/tamagokill/blob/2-ajout-environnents-dev-et-prod-au-docker/api/Dockerfile#L2
You can use the version passed as ARG or define it yourself for further testing

Comment (pbrissaud) [reply web/Dockerfile.dev]: I just set up a build arg `NODE_VERSION` in both dockerfiles to change the node version used in the container. 

The default version is 18. 

You can change it in compose file like that : 

```
build:
  ...
  args:
    NODE_VERSION: 20
```

and the value can be an environment variable.

Comment (pbrissaud) [reply web/Dockerfile.dev]: You can set a NODE_VERSION environment variable to change the node version used to build the image. 

```
web:
  build:
    context: ./web
    dockerfile: Dockerfile.dev
    args:
      - NODE_VERSION=${NODE_VERSION:-18}
```

The default is `18`. You can use any value that matches a node docker tag with a `-alpine` suffix: https://hub.docker.com/_/node/tags?page=&page_size=&ordering=&name=alpine

**Review by ferdodo — APPROVED** [review]

**Review by vasilvestre — APPROVED** [review]

**Review by lilymonad — APPROVED** [review]
